presidio-image-redactor/presidio_image_redactor/dicom_image_pii_verify_engine.py

```diff
@@ -0,0 +1,316 @@
+            matching = [match_left, match_top, match_width, match_height]
+
+            if False not in matching:
+                # If match is found, carry over ground truth info
+                positive = label
+                positive["score"] = results_score
+                all_pos.append(positive)
+                match_found = True
+
+        return all_pos, match_found
+
+    @classmethod
+    def _label_all_positives(
+        cls,
+        gt_labels_dict: dict,
+        ocr_results: List[dict],
+        detected_phi: List[dict],
+        tolerance: int = 50,
+    ) -> List[dict]:
+        """Label all entities detected as PHI by using ground truth and OCR results.
+
+        All positives (detected_phi) do not contain PHI labels and are thus
+        difficult to work with intuitively. This method maps back to the
+        actual PHI to each detected sensitive entity.
+
+        :param gt_labels_dict: Dictionary with ground truth labels for a
+        single DICOM instance.
+        :param ocr_results: All detected text.
+        :param detected_phi: Formatted analyzer_results.
+        :param tolerance: Tolerance for exact coordinates and size data.
+        :return: List of all positives, labeled.
+        """
+        all_pos = []
+
+        # Cycle through each positive (TP or FP)
+        for analyzer_result in detected_phi:
+
+            # See if there are any ground truth matches
+            all_pos, gt_match_found = cls._match_with_source(
+                all_pos, gt_labels_dict, analyzer_result, tolerance
+            )
+
+            # If not, check back with OCR
+            if not gt_match_found:
+                all_pos, _ = cls._match_with_source(
+                    all_pos, ocr_results, analyzer_result, tolerance
+                )
+
+        # Remove any duplicates
+        all_pos = cls._remove_duplicate_entities(all_pos, 0)
+
+        return all_pos
+
+    @staticmethod
+    def calculate_precision(gt: List[dict], all_pos: List[dict]) -> float:
+        """Calculate precision.
+
+        :param gt: List of ground truth labels.
+        :param all_pos: All Detected PHI (mapped back to have actual PHI text).
+        :return: Precision value.
+        """
+        # Find True Positive (TP) and precision
+        tp = [i for i in all_pos if i in gt]
+        try:
+            precision = len(tp) / len(all_pos)
+        except ZeroDivisionError:
+            precision = 0
+
+        return precision
+
+    @staticmethod
+    def calculate_recall(gt: List[dict], all_pos: List[dict]) -> float:
+        """Calculate recall.
+
+        :param gt: List of ground truth labels.
+        :param all_pos: All Detected PHI (mapped back to have actual PHI text).
+        :return: Recall value.
+        """
+        # Find True Positive (TP) and precision
+        tp = [i for i in all_pos if i in gt]
+        try:
+            recall = len(tp) / len(gt)
+        except ZeroDivisionError:
+            recall = 1
```

Comment:
I would go with 0 as default :) but it's hard to say how this would be used by others.